don't delete existing Docker volume if driver unspecified

rhansen opened the pull request:
Before, if the driver parameter wasn't specified, the `dockerng` state module's `volume_present()` function would try to delete the volume and recreate it.  Now it simply ignores the existing volume's driver setting and returns success.

I did not include a test case.  I'll try to find the time to write and push a test case later.

rhansen committed:
don't delete existing Docker volume if driver unspecified

Before, if the driver parameter wasn't specified, the dockerng state
module's volume_present() function would try to delete the volume and
recreate it.  Now it simply ignores the existing volume's driver
setting and returns success.

Note that this change assumes that if the driver is unspecified then
the user means "I don't care what the driver is", not "I care that the
driver matches the Docker default".

ticosax commented on salt/states/dockerng.py:
```diff
@@ -2145,7 +2145,7 @@ def volume_present(name, driver=None, driver_opts=None):
     # volume exits, check if driver is the same.
     volume = volumes[0]
-    if volume['Driver'] != driver:
+    if driver is not None and volume['Driver'] != driver:
```
I think it is better to initiate `driver` parameter with the default docker's volume driver name.

```python
if driver is None:
  driver = get_default_docker_driver()
```
Otherwise when one removes `- driver: my driver` from the sls file, the change to default driver won't be detected.

rhansen replied:
> I think it is better to initiate driver parameter with the default docker's volume driver name.

I disagree—I think it's better to interpret the lack of a specified driver as "I don't care what the driver is" rather than interpret it as "I care very much that the driver matches the Docker default."

(I also think it's a bad idea to delete a volume just because the driver is different.  I think it'd be better to error out than to risk surprising the user with permanent data loss.  But that's best addressed in a different merge request.)

>     driver = get_default_docker_driver()

Does Docker provide a way to get the default driver?  (other than create a temporary volume using defaults, query its properties, then delete it)  If not, I'm not sure how one would implement `get_default_docker_driver()`.  I'd be worried about hard-coding `local` in Salt, as that will cause problems if Docker ever changes the default to something else.

If we can figure out how to cleanly implement `get_default_docker_driver()`, perhaps we can implement a compromise:  `None` would mean "I don't care" while `default` would mean "I care that it matches the Docker default".  (And then hope that Docker never adds a driver named `default` whose semantics differ.)

rhansen committed:
add additional states.dockerng.volume_present() unit tests

The unit tests now test for the bug fixed in the previous commit
(among other things).

rhansen commented:
I pushed a new version of my branch.  Changes:
 
  * rebase onto latest `develop`
  * add a comment to the fix's commit message stating that if the user does not specify a driver then the user means "I don't care what the driver is", not "I care that the driver matches the Docker default"
  * extend `states.dockerng.volume_present()`'s unit tests to test for the bug fixed in this pull request (among other things)

rhansen committed:
document the behavior if the driver is unspecified

rhansen commented:
Go Go Jenkins!

rhansen commented:
The Jenkins test failures seem unrelated to these commits.

ticosax commented:
@rhansen Thanks to the docstring explaining what it does regarding default Driver, I think your approach is good to go.

thanks for the additional tests.

:+1:

rhansen commented:
Thank you both!
